src/lib/provable/bigint.ts

```diff
@@ -0,0 +1,898 @@
+ * arithmetic operations to maintain valid representations. All operations are
+ * designed to be provable and optimised for less constraints.
+ *
+ * @param modulus The modulus for the big integer arithmetic (must be prime)
+ * @param config Optional configuration specifying a custom limb size and number
+ * @returns A class representing ProvableBigInts with the specified modulus
+ * @throws If the modulus is zero, negative, or exceeds the maximum supported size
+ */
+function createProvableBigInt(modulus: bigint, config?: BigIntParameter) {
+  const config_ = config ?? findConfig(modulus);
+  assert(modulus !== 0n, `ProvableBigInt: modulus must be non-zero, got ${modulus}`);
+  assert(modulus > 0n, `ProvableBigInt: modulus must be positive, got ${modulus}`);
+  assert(
+    modulus < config_.MAX,
+    `ProvableBigInt: modulus exceeds the max supported size of 2^${config_.MAX}`
+  );
+  let fields = [];
+  let x = modulus;
+  if (x < 0n) {
+    throw new Error('Modulus must be non-negative.');
+  }
+  if (x > config_.MAX) {
+    throw new Error(
+      `Modulus exceeds ${config_.limb_num * Number(config_.limb_size)}-bit size limit.`
+    );
+  }
+
+  for (let i = 0; i < config_.limb_num; i++) {
+    fields.push(Field.from(x & config_.mask)); // fields[i] = x & mask
+    x >>= config_.limb_size; // x = x >> limb_size
+  }
+
+  class ProvableBigInt_ extends ProvableBigInt<ProvableBigInt_> {
+    constructor(fields: Field[], value: Unconstrained<bigint>) {
+      super(fields, Unconstrained.from(value));
+    }
+
+    static {
+      this._modulus = new ProvableBigInt_(fields, Unconstrained.from(modulus));
+      this._config = config_;
+      this._provable = provableFromClass(ProvableBigInt_, {
+        fields: Provable.Array(Field, config_.limb_num),
+      });
+    }
+
+    /**
+     * Returns a ProvableBigInt representing zero
+     * @returns A ProvableBigInt representing zero
+     */
+    static zero(): ProvableBigInt_ {
+      return ProvableBigInt_.fromBigInt(0n);
+    }
+
+    /**
+     * Returns a ProvableBigInt representing one
+     * @returns A ProvableBigInt representing one
+     */
+    static one(): ProvableBigInt_ {
+      return ProvableBigInt_.fromBigInt(1n);
+    }
+
+    /**
+     * Returns a ProvableBigInt representing one
+     * @returns A ProvableBigInt representing one
+     */
+    static MAX(): ProvableBigInt_ {
+      return ProvableBigInt_.fromBigInt(modulus - 1n);
+    }
+
+    static from(x: bigint | string | number | boolean): ProvableBigInt_ {
+      return ProvableBigInt_.fromBigInt(BigInt(x));
+    }
+
+    static fromBigInt(x: bigint): ProvableBigInt_ {
+      let fields = [];
+      let value = x;
+
+      if (value < 0n) {
+        value = ((x % modulus) + modulus) % modulus;
+      }
+      if (value >= ProvableBigInt_.modulus.toBigInt()) {
+        value = value % modulus;
+      }
+      for (let i = 0; i < ProvableBigInt_.config.limb_num; i++) {
```

Comment:
Seems like this is quite similar to what's on line [142](https://github.com/o1-labs/o1js/blob/5b87a1aa69bafde33b7e2139c3d6b4348b23a38c/src/lib/provable/bigint.ts#L142-L145), I wonder if it makes sense to make this modular to avoid code duplication.